source/cpu_usage_sampler.c

```diff
@@ -0,0 +1,238 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/cpu_usage_sampler.h>
+
+#if defined(__linux__) || defined(__unix__)
+#    include <sys/sysinfo.h>
+#    include <sys/types.h>
+#endif
+
+/*********************************************************************************************************************
+ * Base implementation
+ ********************************************************************************************************************/
+
+struct aws_cpu_sampler_vtable {
+    int (*aws_get_cpu_sample_fn)(struct aws_cpu_sampler *sampler, double *output);
+    void (*aws_cpu_sampler_destroy)(struct aws_cpu_sampler *sampler);
+};
+
+struct aws_cpu_sampler {
+    const struct aws_cpu_sampler_vtable *vtable;
+    struct aws_allocator *allocator;
+    void *impl;
+};
+
+static void aws_cpu_sampler_destroy_default(struct aws_cpu_sampler *sampler) {
+    if (sampler == NULL) {
+        return;
+    }
+    aws_mem_release(sampler->allocator, sampler);
+}
+
+static struct aws_cpu_sampler_vtable aws_cpu_sampler_vtable_default = {
+    .aws_get_cpu_sample_fn = NULL,
+    .aws_cpu_sampler_destroy = aws_cpu_sampler_destroy_default,
+};
+
+/*********************************************************************************************************************
```

Comment:
You're on the right track in terms of data structures.  The missing component is that you want implementations in each of source/windows/ (empty, do nothing), source/darwin (empty, do nothing), and source/linux (implementation with no conditional compilation).  Unfortunately, we don't have darwin and linux folders in aws-c-common the same way we do in aws-c-io.  Seeing what people think right now.